gh-84538: add strict argument to pathlib.PurePath.relative_to (#19813)
python/cpython: Doc/library/pathlib.rst

      >>> p.relative_to('/usr')
      Traceback (most recent call last):
        File "<stdin>", line 1, in <module>
        File "pathlib.py", line 941, in relative_to
          raise ValueError(error_message.format(str(self), str(formatted)))
      ValueError: '/etc/passwd' is not in the subpath of '/usr' OR one path is relative and the other is absolute.
      >>> p.relative_to('/usr', strict=False)
      PurePosixPath('../etc/passwd')
      >>> p.relative_to('foo', strict=False)
      Traceback (most recent call last):
        File "<stdin>", line 1, in <module>
        File "pathlib.py", line 941, in relative_to
          raise ValueError(error_message.format(str(self), str(formatted)))
      ValueError: '/etc/passwd' is not on the same drive as 'foo' OR one path is relative and the other is absolute.

   If the path doesn't start with *other* and *strict* is ``True``, :exc:`ValueError` is raised.  If *strict* is ``False`` and one path is relative and the other is absolute or if they reference different drives :exc:`ValueError` is raised.

   NOTE: This function is part of :class:`PurePath` and works with strings. It does not check or access the underlying file structure.

>    .. versionadded:: 3.10
>       The *strict* argument (pre-3.10 behavior is strict).

Reviewer: ```suggestion
   .. versionadded:: 3.11
      The *strict* argument (pre-3.11 behavior is strict).
```

      >>> p.relative_to('/usr')
      Traceback (most recent call last):
        File "<stdin>", line 1, in <module>
        File "pathlib.py", line 941, in relative_to
          raise ValueError(error_message.format(str(self), str(formatted)))
      ValueError: '/etc/passwd' is not in the subpath of '/usr' OR one path is relative and the other is absolute.
      >>> p.relative_to('/usr', strict=False)
      PurePosixPath('../etc/passwd')
      >>> p.relative_to('foo', strict=False)
      Traceback (most recent call last):
        File "<stdin>", line 1, in <module>
        File "pathlib.py", line 941, in relative_to
          raise ValueError(error_message.format(str(self), str(formatted)))
      ValueError: '/etc/passwd' is not on the same drive as 'foo' OR one path is relative and the other is absolute.

>    If the path doesn't start with *other* and *strict* is ``True``, :exc:`ValueError` is raised.  If *strict* is ``False`` and one path is relative and the other is absolute or if they reference different drives :exc:`ValueError` is raised.

Reviewer 2: I think the key difference should be spelled out in terms of whether adding `..` entries is allowed. How about something like this?

```
In strict mode (the default), the path must start with *other*. In non-strict 
mode, ``..`` entries may be added to form the relative path. In all other 
cases, such as the paths referencing different drives, :exe:`ValueError` is 
raised.

.. warning::
    Non-strict mode assumes that no symlinks are present in the path; you 
    should call :meth:`~Path.resolve` first to ensure this.
```

   .. versionadded:: 3.11
      The *strict* argument (pre-3.11 behavior is strict).
   .. versionadded:: 3.10

>       The *strict* argument (pre-3.10 behavior is strict).

Reviewer 2: Duplication?

Author: Oh, I didn't notice this, thanks.

      >>> p.relative_to('/usr')
      Traceback (most recent call last):
        File "<stdin>", line 1, in <module>
        File "pathlib.py", line 941, in relative_to
          raise ValueError(error_message.format(str(self), str(formatted)))
      ValueError: '/etc/passwd' is not in the subpath of '/usr' OR one path is relative and the other is absolute.
      >>> p.relative_to('/usr', walk_up=True)
      PurePosixPath('../etc/passwd')
      >>> p.relative_to('foo', walk_up=True)
      Traceback (most recent call last):
        File "<stdin>", line 1, in <module>
        File "pathlib.py", line 941, in relative_to
          raise ValueError(error_message.format(str(self), str(formatted)))
      ValueError: '/etc/passwd' is not on the same drive as 'foo' OR one path is relative and the other is absolute.

   When *walk_up* is False, the default, the path must start with *other*,
   when it's True ``..`` entries may be added to form the relative path. In
   all other cases, such as the paths referencing different drives,

>    :exc:`ValueError` is raised.

Reviewer 2: I think this documentation needs to occur *before* the examples of using walk_up.

So, either move it after "ValueError is raised", or split the examples into two parts.

      >>> p.relative_to('/usr')
      Traceback (most recent call last):
        File "<stdin>", line 1, in <module>
        File "pathlib.py", line 941, in relative_to
          raise ValueError(error_message.format(str(self), str(formatted)))
      ValueError: '/etc/passwd' is not in the subpath of '/usr' OR one path is relative and the other is absolute.

>       When *walk_up* is False (the default), the path must start with *other*.

Reviewer 3: ```suggestion
When *walk_up* is False (the default), the path must start with *other*.
```

The warning is due to this accidentally being indented, which includes it in the code literal block

   when it's True ``..`` entries may be added to form the relative path. In
   all other cases, such as the paths referencing different drives,

>    :exc:`ValueError` is raised.::

Reviewer 4: ```suggestion
      When *walk_up* is False (the default), the path must start with *other*.
   When the argument is True, ``..`` entries may be added to form the
   relative path. In all other cases, such as the paths referencing
   different drives, :exc:`ValueError` is raised.::
```

      >>> p.relative_to('/usr', walk_up=True)
      PurePosixPath('../etc/passwd')
      >>> p.relative_to('foo', walk_up=True)
      Traceback (most recent call last):
        File "<stdin>", line 1, in <module>
        File "pathlib.py", line 941, in relative_to
          raise ValueError(error_message.format(str(self), str(formatted)))
      ValueError: '/etc/passwd' is not on the same drive as 'foo' OR one path is relative and the other is absolute.

   .. warning::
      The *walk_up* option assumes that no symlinks are present in the path; you

>       should call :meth:`~Path.resolve` first to ensure this.

Reviewer 4: ```suggestion
      The *walk_up* option assumes that no symlinks are present in the path;
      call :meth:`~Path.resolve` first if necessary to resolve symlinks.
```

>    .. warning::
>       The *walk_up* option assumes that no symlinks are present in the path;
>       call :meth:`~Path.resolve` first if necessary to resolve symlinks.

Reviewer 4: ```suggestion
   .. warning::
      This function is part of :class:`PurePath` and works with strings.
      It does not check or access the underlying file structure.
      This can impact the *walk_up* option as it assumes that no symlinks
      are present in the path; call :meth:`~Path.resolve` first if
      necessary to resolve symlinks.
```

   .. versionadded:: 3.12

>       The *walk_up* argument (pre-3.12 behavior is the same as walk_up=False).

Reviewer 4: ```suggestion
      The *walk_up* argument (old behavior is the same as ``walk_up=False``).
```
